File: macos/Sources/ElevateCore/Providers/EntraApprovalProvider.swift
import Foundation

/// Entra directory role activation requests awaiting the signed-in user's approval.
public struct EntraApprovalProvider: ApprovalProvider {
    public let kind: RoleScopeKind = .entraDirectory
    public let scopes = GraphScopes.all
    let transport: GraphTransport

    public init(http: any HTTPClient, tokens: any TokenProviding) {
        transport = GraphTransport(http: http, tokens: tokens)
    }

    struct Request: Decodable {
        let id: String
        let action: String?
        let principalId: String?
        let roleDefinitionId: String?
        let justification: String?
        let createdDateTime: Date?
        let scheduleInfo: GraphApprovals.ScheduleInfo?
        let roleDefinition: GraphApprovals.Named?
        let principal: GraphApprovals.Principal?
    }

    static let approvalsPath = "/roleManagement/directory/roleAssignmentApprovals"

    public func pendingApprovals(identity: Identity, tenant: TenantContext) async throws -> [ApprovalRequest] {
        let url = try transport.graphURL("/roleManagement/directory/roleAssignmentScheduleRequests/filterByCurrentUser(on='approver')?$filter=status eq 'PendingApproval'&$expand=roleDefinition,principal")
        let items = try await transport.listAll(Request.self, identity: identity, tenantId: tenant.tenantId, url: url, scopes: scopes)
        return items.map { r in
            ApprovalRequest(id: r.id, tenantKey: tenant.id, kind: kind, action: GraphApprovals.action(r.action),
                            targetName: r.roleDefinition?.displayName ?? r.roleDefinitionId ?? r.id,
                            requesterName: GraphApprovals.requester(r.principal, principalId: r.principalId),
                            justification: r.justification,
                            requestedDuration: r.scheduleInfo?.expiration?.duration.flatMap(ISO8601Duration.parse),
                            createdAt: r.createdDateTime, decisionRef: r.id)
        }
    }

    public func decide(_ request: ApprovalRequest, approve: Bool, justification: String, identity: Identity) async throws {
        try await GraphApprovals.decide(request, approve: approve, justification: justification, identity: identity,
                                        transport: transport, scopes: scopes, approvalsPath: Self.approvalsPath)
    }
}

File: windows/src/Elevate.Core/Providers/EntraApprovalProvider.cs
using Elevate.Core.Auth;
using Elevate.Core.Models;
using Elevate.Core.Networking;

namespace Elevate.Core.Providers;

/// <summary>Entra directory role activation requests awaiting the signed-in user's approval.</summary>
public sealed class EntraApprovalProvider : IApprovalProvider
{
    internal const string ApprovalsPath = "/roleManagement/directory/roleAssignmentApprovals";

    private const string PendingPath =
        "/roleManagement/directory/roleAssignmentScheduleRequests/filterByCurrentUser(on='approver')"
        + "?$filter=status eq 'PendingApproval'&$expand=roleDefinition,principal";

    private readonly GraphTransport _transport;

    public EntraApprovalProvider(IHttpClient http, ITokenProvider tokens)
        => _transport = new GraphTransport(http, tokens);

    public RoleScopeKind Kind => RoleScopeKind.EntraDirectory;

    public IReadOnlyList<string> Scopes => Auth.Scopes.GraphAll;

    private sealed record Request(
        string Id,
        string? Action,
        string? PrincipalId,
        string? RoleDefinitionId,
        string? Justification,
        DateTimeOffset? CreatedDateTime,
        ScheduleInfo? ScheduleInfo,
        GraphApprovals.Named? RoleDefinition,
        GraphApprovals.Principal? Principal);

    public async Task<IReadOnlyList<ApprovalRequest>> PendingApprovalsAsync(
        Identity identity, TenantContext tenant, CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(identity);
        ArgumentNullException.ThrowIfNull(tenant);

        var items = await _transport.ListAllAsync<Request>(
            identity, tenant.TenantId, _transport.GraphUrl(PendingPath), Scopes, ct).ConfigureAwait(false);

        return [.. items.Select(r => new ApprovalRequest(
            r.Id, tenant.Key, Kind, GraphApprovals.Action(r.Action),
            targetName: r.RoleDefinition?.DisplayName ?? r.RoleDefinitionId ?? r.Id,
            requesterName: GraphApprovals.Requester(r.Principal, r.PrincipalId),
            justification: r.Justification,
            requestedDuration: GraphApprovals.Duration(r.ScheduleInfo),
            createdAt: r.CreatedDateTime,
            decisionRef: r.Id))];
    }

    public Task DecideAsync(ApprovalRequest request, bool approve, string justification, Identity identity, CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(request);
        return GraphApprovals.DecideAsync(request, approve, justification, identity, _transport, Scopes, ApprovalsPath, ct);
    }
}
